vms/evm/predicate/predicate_test.go

```diff
@@ -0,0 +1,278 @@
+		})
+	}
+}
+
+func TestPredicateBytes(t *testing.T) {
+	tests := []struct {
+		name    string
+		input   Predicate
+		want    []byte
+		wantErr error
+	}{
+		// Valid test cases
+		{
+			name:    "empty_input",
+			input:   New(nil),
+			want:    []byte{},
+			wantErr: nil,
+		},
+		{
+			name:    "single_byte",
+			input:   New([]byte{0xbb}),
+			want:    []byte{0xbb},
+			wantErr: nil,
+		},
+		{
+			name:    "31_bytes",
+			input:   New(bytes.Repeat([]byte{0xaa}, 31)),
+			want:    bytes.Repeat([]byte{0xaa}, 31),
+			wantErr: nil,
+		},
+		{
+			name:    "32_bytes",
+			input:   New(bytes.Repeat([]byte{0xdd}, 32)),
+			want:    bytes.Repeat([]byte{0xdd}, 32),
+			wantErr: nil,
+		},
+		{
+			name:    "33_bytes",
+			input:   New(bytes.Repeat([]byte{0xcc}, 33)),
+			want:    bytes.Repeat([]byte{0xcc}, 33),
+			wantErr: nil,
+		},
+		{
+			name:    "48_bytes",
+			input:   New(bytes.Repeat([]byte{0x00}, 48)),
+			want:    bytes.Repeat([]byte{0x00}, 48),
+			wantErr: nil,
+		},
+		{
+			name:    "63_bytes",
+			input:   New(bytes.Repeat([]byte{0xdd}, 63)),
+			want:    bytes.Repeat([]byte{0xdd}, 63),
+			wantErr: nil,
+		},
+		{
+			name:    "64_bytes",
+			input:   New(bytes.Repeat([]byte{0x33}, 64)),
+			want:    bytes.Repeat([]byte{0x33}, 64),
+			wantErr: nil,
+		},
+		{
+			name:    "65_bytes",
+			input:   New(bytes.Repeat([]byte{0xdd}, 65)),
+			want:    bytes.Repeat([]byte{0xdd}, 65),
+			wantErr: nil,
+		},
+		// Invalid test cases
+		{
+			name:    "all_zeros_empty",
+			input:   nil,
+			want:    nil,
+			wantErr: errMissingDelimiter,
+		},
+		{
+			name:    "all_zeros",
+			input:   Predicate{{}},
+			want:    nil,
+			wantErr: errMissingDelimiter,
+		},
+		{
+			name:    "wrong_delimiter",
+			input:   Predicate{{0x42}},
+			want:    nil,
+			wantErr: errWrongDelimiter,
+		},
+		{
+			name:    "wrong_delimiter",
+			input:   Predicate{{delimiter}, {}},
+			want:    nil,
+			wantErr: errExcessPadding,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require := require.New(t)
+
+			unpacked, err := tt.input.Bytes()
+			require.ErrorIs(err, tt.wantErr)
+			require.Equal(tt.want, unpacked)
+		})
+	}
+}
+
+func FuzzNewBytesEqual(f *testing.F) {
+	f.Fuzz(func(t *testing.T, input []byte) {
+		packed := New(input)
+		unpacked, err := packed.Bytes()
+		require.NoError(t, err)
+		require.Equal(t, input, unpacked)
+	})
+}
+
+func FuzzBytesNewEqual(f *testing.F) {
+	f.Fuzz(func(t *testing.T, original []byte) {
+		var predicate Predicate
+		for i := 0; i < len(original); i += common.HashLength {
+			var key common.Hash
+			copy(key[:], original[i:])
+			predicate = append(predicate, key)
+		}
+
+		unpacked, err := predicate.Bytes()
+		if err != nil {
+			t.Skip("invalid predicate")
+		}
+
+		repacked := New(unpacked)
+		require.Equal(t, predicate, repacked)
+	})
+}
+
+type allowSet map[common.Address]bool
+
+func (a allowSet) HasPredicate(addr common.Address) bool { return a[addr] }
```

Comment:
Seems odd that we use a `bool` as the value in this map since we only care if the key is/isn't present - the value is always supposed to be true for our usage pattern.

```suggestion
type allowSet set.Set[common.Address]

func (a allowSet) HasPredicate(addr common.Address) bool { return a.Contains(addr) }
```

Reply:
Oh. We could just implement it as a struct with a set field then. A `map[T]struct{}` is a set anyways.

Reply:
I suspect using `set.Of` would alleviate the concerns about cleanliness